workflow/controller/operator.go

```diff
@@ -3074,76 +3090,99 @@ func (woc *wfOperationCtx) fetchWorkflowSpec() (*wfv1.WorkflowSpec, error) {
 	if err != nil {
 		return nil, err
 	}
-	return specHolder.GetWorkflowSpec(), nil
+	return specHolder, nil
 }
 
-func (woc *wfOperationCtx) loadExecutionSpec() (wfv1.TemplateReferenceHolder, wfv1.Arguments, error) {
-
-	executionParameters := woc.wf.Spec.Arguments
-
-	if woc.wf.Spec.WorkflowTemplateRef == nil {
-		if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
-			return nil, executionParameters, fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
-		}
-		// Set the WorkflowDefaults from Configmap
-		err := woc.controller.setWorkflowDefaults(woc.wf)
-		if err != nil {
-			log.WithFields(log.Fields{"key": woc.wf.Name, "error": err}).Warn("Failed to apply default workflow values")
-			return nil, executionParameters, err
-		}
-
-		tmplRef := &wfv1.WorkflowStep{Template: woc.wf.Spec.Entrypoint}
-		return tmplRef, executionParameters, nil
+func (woc *wfOperationCtx) retryStrategy(tmpl *wfv1.Template) *wfv1.RetryStrategy {
+	if tmpl != nil && tmpl.RetryStrategy != nil {
+		return tmpl.RetryStrategy
 	}
+	return woc.execWf.Spec.RetryStrategy
+}
 
-	if woc.wf.Status.StoredWorkflowSpec == nil {
-		wftSpec, err := woc.fetchWorkflowSpec()
+func (woc *wfOperationCtx) loadExecWFFromWfDefaultOrWFTRef() error {
+	if woc.wf.Spec.WorkflowTemplateRef != nil {
+		err := woc.setStoredWFSpec()
 		if err != nil {
-			return nil, executionParameters, err
+			return err
 		}
-		woc.wf.Status.StoredWorkflowSpec = wftSpec
-		woc.updated = true
-	} else if woc.controller.Config.WorkflowRestrictions.MustNotChangeSpec() {
-		// If the controller is in reference mode, ensure that the stored spec is identical to the reference spec at every operation
-		wftSpec, err := woc.fetchWorkflowSpec()
+		err = woc.setExecWorkflow()
 		if err != nil {
-			return nil, executionParameters, err
+			return err
 		}
-		if woc.wf.Status.StoredWorkflowSpec.String() != wftSpec.String() {
-			return nil, executionParameters, fmt.Errorf("workflowTemplateRef reference may not change during execution when the controller is in reference mode")
+	} else if woc.controller.Config.WorkflowRestrictions.MustUseReference() {
+		return fmt.Errorf("workflows must use workflowTemplateRef to be executed when the controller is in reference mode")
+	} else {
+		err := woc.controller.setWorkflowDefaults(woc.wf)
+		if err != nil {
+			return err
 		}
+		woc.volumes = woc.wf.Spec.DeepCopy().Volumes
 	}
+	return nil
+}
+
+func (woc *wfOperationCtx) setStoredWFSpec() error {
```

Comment:
This function is very messy and disorganized: it's impossible to parse and its convoluted structure is likely to introduce bugs. Can you please spend some time tidying it up? It's a crucial piece of code, so it's worth investing time making sure it's readable and well-structured.

Reply:
By the way: the function is now much easier to read. Thanks.